src/core/reactor_backend.cc

```diff
@@ -882,6 +884,18 @@ reactor_backend_epoll::wait_and_process(int timeout, const sigset_t* active_sigm
             auto op = evt.events ? EPOLL_CTL_MOD : EPOLL_CTL_DEL;
             ::epoll_ctl(_epollfd.get(), op, pfd->fd.get(), &evt);
         }
+        else if (has_error && pfd->events_requested == 0) {
+            // In case we got an error and we are not interested in any events
+            // anymore, remove the fd from the epoll set. Otherwise due to the
+            // above logic of setting evt.events to pfd->events_requested (0)
+            // and subsequent calculation of events_to_remove, we would never
+            // remove the fd from the epoll set. Due to running in level
+            // triggered mode this would then cause repeated wake ups and
+            // spinning until the fd is closed and thereby removed from the
+            // epoll set.
+            pfd->events_epoll = 0;
+            ::epoll_ctl(_epollfd.get(), EPOLL_CTL_DEL, pfd->fd.get(), nullptr);
```

Comment:
This affects both the read end and the write end. What if the write end is still open?

Perhaps we should add EPOLLIN to events_to_remove and let the if (events_to_remove) branch choose between EPOLL_CTL_MOD and EPOLL_CTL_ADD.

Reply:
> This affects both the read end and the write end. What if the write end is still open?

By write end do you actually mean our write site i.e.: a possible registration of EPOLLOUT or the remote end?

Re the former, as currently implemented we would only take this path if we are not interested in any events anymore because of the `pfd->events_requested == 0` check in the if.

However, as per the other comment given how the whole handling is currently implemented the moment we receive an error we will remove everything from `pfd->events_requested` anyway and hence would ignore any further events in any case.

Re the later, I don't think that should matter in that regard. As stated above if we received any further events then we would not get notified of them anymore. However given `events_requested` is 0 now even if we got notified we would not deliver them anywhere.

Want to state that this can happen on both EPOLLIN and EPOLLOUT but that doesn't really change your point or what's written above (just that any handling specific to either of the two would have to handle both cases).

Reply:
> > This affects both the read end and the write end. What if the write end is still open?
> 
> By write end do you actually mean our write site i.e.: a possible registration of EPOLLOUT or the remote end?


A possible registration of EPOLLOUT.

> 
> Re the former, as currently implemented we would only take this path if we are not interested in any events anymore because of the `pfd->events_requested == 0` check in the if.

Having re-read the code I agree, but from a higher-level point of view: the code prefers to generate a spurious wakeup (complete_epoll_event()). If the wakeup was wanted, great. If not wanted, the waiter will re-register.


> 
> However, as per the other comment given how the whole handling is currently implemented the moment we receive an error we will remove everything from `pfd->events_requested` anyway and hence would ignore any further events in any case.
> 
> Re the later, I don't think that should matter in that regard. As stated above if we received any further events then we would not get notified of them anymore. However given `events_requested` is 0 now even if we got notified we would not deliver them anywhere.
> 
> Want to state that this can happen on both EPOLLIN and EPOLLOUT but that doesn't really change your point or what's written above (just that any handling specific to either of the two would have to handle both cases).

Yes. What about reducing complexity by using events_to_remove to de-register? Right now both has_error and events_to_remove do the same thing.

Reply:
> Yes. What about reducing complexity by using events_to_remove to de-register? Right now both has_error and events_to_remove do the same thing.

I have pushed a change which uses events_to_remove. 

I am not sure it removes complexity but I'll let you guys be the judge. As touched on in the other comment I originally tried to basically leave the existing code as is and just add additional handling so as not to infer with the already existing complexity of how events and events_to_remove is calculated.

However this change should work equally well. If you prefer the new version I will squash the commits.

cc @michoecho

Reply:
I think it's simpler. Could be simplified a bit further by calculating events_to_remove early, then overriding it the error branch, but we can leave it for later. I prefer not having a maze of conditionals.

Please squash the commits.

Reply:
Commits are squashed btw